### PR TITLE
Add scrollable trait to the ListEditor

### PR DESCRIPTION
**Pull request by mayankmanj**
The `scrollable` trait of the `ListEditor` allows us to hide the vertical scrollbar, and display all the items contained in the list.

Drive-by fixes -- Removes some duplicate code.

### CODE REVIEW

File: traitsui/qt4/list_editor.py [hunk 1]
```diff
@@ -382,6 +382,21 @@ def _mutable_default(self):
         """
         return self.factory.mutable
 
+    def _scrollable_default(self):
+        """ Trait handler to set the scrollable from the factory.
+        """
+        return self.factory.scrollable
+
+    def __list_pane_default(self):
```
Comment (corranwebster) [inline traitsui/qt4/list_editor.py]: I'd just create this directly in the init; or if you want to encapsulate, have this as a private method which creates the `list_pane` attribute.

File: traitsui/qt4/list_editor.py [hunk 2]
```diff
@@ -742,21 +750,3 @@ def _menu_action(self, event, name=""):
         else:
             # TODO: Fix tab order based on the context_object's list
             self.control.addTab(self._pagewidgets[name], name)
-
-    def _context_menu_requested(self, event):
```
Comment (corranwebster) [inline traitsui/qt4/list_editor.py]: Why is this removed?  This seems unrelated to the main change?

File: traitsui/qt4/list_editor.py [hunk 3]
```diff
@@ -382,6 +382,21 @@ def _mutable_default(self):
         """
         return self.factory.mutable
 
+    def _scrollable_default(self):
```
Comment (corranwebster) [inline traitsui/qt4/list_editor.py]: You don't have a `scrollable` trait declared on the list editor (as opposed to the list editor factory).  So this _probably_ isn't working the way that you expect.

File: traitsui/qt4/list_editor.py [hunk 4]
```diff
@@ -221,10 +222,9 @@ def update_editor_item(self, event):
 
         # Otherwise, find the proxy for this index and update it with the
         # changed value:
-        for control in self.control.widget().children():
-            if isinstance(control, QtGui.QLayout):
-                continue
-
+        children = (self._layout.itemAt(ind).widget()
+                    for ind in range(self._layout.count()))
+        for control in children:
```
Comment (corranwebster) [inline traitsui/qt4/list_editor.py]: A better way to do this would just be:
```
for control in self._list_pane.children():
```
I think.  Then you don't need to track the `_layout` specifically as a trait.

File: traitsui/qt4/list_editor.py [hunk 5]
```diff
@@ -96,20 +98,19 @@ def init(self, parent):
...
         # Create a mapper to identify which icon button requested a contextmenu
         self.mapper = QtCore.QSignalMapper(self.control)
 
         # Create a widget with a grid layout as the container.
-        self._list_pane = QtGui.QWidget()
-        self._list_pane.setSizePolicy(
-            QtGui.QSizePolicy.Expanding, QtGui.QSizePolicy.Expanding
-        )
-        layout = QtGui.QGridLayout(self._list_pane)
+        self._layout = layout = QtGui.QGridLayout(self._list_pane)
```
Comment (corranwebster) [inline traitsui/qt4/list_editor.py]: `_layout` isn't declared as a trait on the class (but I don't think you actually need it - see previous comment).

File: traitsui/editors/list_editor.py [hunk 6]
```diff
@@ -85,6 +85,9 @@ class ToolkitEditorFactory(EditorFactory):
     #: Can the list be reorganized, or have items added and deleted.
     mutable = Bool(True)
 
+    #: Should the scrollbars be displayed if the list is too long.
+    scrollable = Bool(True)
```
Comment (corranwebster) [inline traitsui/editors/list_editor.py]: If you declare as:
```
scrollable = Bool(True, sync_value=True)
```
then the value will be copied from the factory to the actual editor automatically.

This is a new-ish feature.

Comment (mayankmanj) [reply traitsui/editors/list_editor.py]: This is not working for me -- the value is not being copied from the factory to the editor. I have defined a `scrollable` trait in the editor (set to `True` by default), and removed the `_scrollable_default` method that was copying the value earlier.

File: traitsui/qt4/list_editor.py (hunk 3 above)
Comment (mayankmanj) [reply traitsui/qt4/list_editor.py]: The `scrollable` was present in the parent class `UIEditor` -- but I realize now that it is set to `False` by default. Setting default to `True` here.

File: traitsui/qt4/list_editor.py (hunk 2 above)
Comment (mayankmanj) [reply traitsui/qt4/list_editor.py]: This is duplicate code from lines [728-744](https://github.com/enthought/traitsui/blob/35f0510ca3090574f121786221548b09ac1882a8/traitsui/qt4/list_editor.py#L728-L744) above.

Would you like me to address this in a separate PR instead?

File: docs/source/traitsui_user_manual/factories_basic.rst [hunk 7]
```diff
@@ -923,11 +923,13 @@ reordering items within the list.
 The simple style displays a single item at a time, with small arrows on the
 right side to scroll the display. The custom style shows multiple items. The
 number of items displayed is controlled by the *rows* parameter; if the number
```
Comment (mayankmanj) [inline docs/source/traitsui_user_manual/factories_basic.rst]: I think the `rows` parameter is used only in WX, not in QT. Should this be mentioned here?

Comment (mayankmanj) [reply docs/source/traitsui_user_manual/factories_basic.rst]: Maybe a better solution for controlling the display of scrollbar is simply to respect the `rows` parameter in `QT`.

File: traitsui/qt4/list_editor.py (hunk 2 above)
Comment (corranwebster) [reply traitsui/qt4/list_editor.py]: Ah.  That makes sense - it wasn't obvious from the diff.  It's fine to do drive-by fixes like this, just mention it in the PR description at the top that you are also doing this change.

File: docs/source/traitsui_user_manual/factories_basic.rst (hunk 7 above)
Comment (corranwebster) [reply docs/source/traitsui_user_manual/factories_basic.rst]: Rows is meant to be a vertical size hint.  I can see the temptation to try and use it, but we don't support changing it dynamically at the moment, which is what would be needed for removing scrollability in the general case.  Something like the current PR is a better approach.

It wouldn't be a bad idea to be able to support the rows parameter in the Qt backend, but that should probably be a different PR (and obviously would be ignored if Scrollable is False.

File: traitsui/editors/list_editor.py (hunk 6 above)
Comment (corranwebster) [reply traitsui/editors/list_editor.py]: Hmmm... possibly a bug somewhere, then, but this line should copy things from the factory to the editor: https://github.com/enthought/traitsui/blob/master/traitsui/editor.py#L584

**Comment by mayankmanj** [comment]
Thank you! Merging.
